builtin/virtualization.c

```diff
@@ -0,0 +1,102 @@
+#include "builtin.h"
+#include "config.h"
+#include "json-writer.h"
+#include "parse-options.h"
+#include "read-cache-ll.h"
+#include "repository.h"
+#include "trace2.h"
+
+static const char * const virt_usage[] = {
+	N_("(EXPERIMENTAL!) git virtualization"),
+	NULL,
+};
+
+static struct option virt_options[] = {
+	OPT_END(),
+};
+
+static int virt_load_config_cb(const char *var,
+			       const char *value,
+			       const struct config_context *ctx,
+			       void *pvoid)
+{
+	return git_default_config(var, value, ctx, pvoid);
+}
+
+static void virt_load_config(void)
+{
+	git_config(virt_load_config_cb, NULL);
+}
+
+static uint64_t count_skipped(struct repository *repo)
+{
+	uint64_t k;
+	uint64_t count = 0;
+
+	for (k = 0; k < repo->index->cache_nr; k++) {
+		const struct cache_entry *ce_k = repo->index->cache[k];
+
+		if (ce_skip_worktree(ce_k))
+			count++;
+	}
+
+	return count;
+}
+
+static void print_hydration(struct repository *repo)
+{
+	uint64_t c_skipped;
+	uint64_t c_total;
+	uint64_t c_hydrated;
+	double hydration;
+
+	c_skipped = count_skipped(the_repository);
+	c_total = (uint64_t)the_repository->index->cache_nr;
+	c_hydrated = c_total - c_skipped;
```

Comment:
So this is really measuring "how many paths in the index have the skip-wortree bit" and using that to measure non-hydrated paths. This doesn't need to be specific to virtual paths.

You could implement this as a subcommand of `git sparse-checkout`, say `git sparse-checkout stats`. Bonus points if you are sparse index aware and report the density of skip-worktree _directories_ before expanding the index and counting how many skip-worktree _files_ would exist otherwise.

Reply:
I was wondering about also having sparse-index stats.  It would be cool to have pre- and post-expanded stats.  I was mainly thinking about solving a problem for GVFS support staff, but there may be a wider audience for it.